pkg/restic/backupper_restorer.go

```diff
@@ -0,0 +1,286 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restic
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	corev1api "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kerrs "k8s.io/apimachinery/pkg/util/errors"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	arkv1api "github.com/heptio/ark/pkg/apis/ark/v1"
+	"github.com/heptio/ark/pkg/util/kube"
+)
+
+type backupperRestorer struct {
+	metadataManager   RepositoryManager
+	daemonSetExecutor DaemonSetExecutor
+	pvcGetter         corev1client.PersistentVolumeClaimsGetter
+}
+
+func NewBackupperRestorer(
+	metadataManager RepositoryManager,
+	daemonSetExecutor DaemonSetExecutor,
+	pvcGetter corev1client.PersistentVolumeClaimsGetter,
+) BackupperRestorer {
+	return &backupperRestorer{
+		metadataManager:   metadataManager,
+		daemonSetExecutor: daemonSetExecutor,
+		pvcGetter:         pvcGetter,
+	}
+}
+
+type BackupperRestorer interface {
+	Backupper
+	Restorer
+}
+
+// Backupper can execute restic backups of volumes in a pod.
+type Backupper interface {
+	BackupPodVolumes(backup *arkv1api.Backup, pod *corev1api.Pod, log logrus.FieldLogger) error
+}
+
+// Restorer can execute restic restores of volumes in a pod.
+type Restorer interface {
+	RestorePodVolumes(restore *arkv1api.Restore, pod *corev1api.Pod, log logrus.FieldLogger) error
+}
+
+func (br *backupperRestorer) BackupPodVolumes(backup *arkv1api.Backup, pod *corev1api.Pod, log logrus.FieldLogger) error {
+	// get volumes to backup from pod's annotations
+	volumesToBackup := GetVolumesToBackup(pod)
+	if len(volumesToBackup) == 0 {
+		return nil
+	}
+
+	// Get existing snapshots annotated on backup, and fail-fast if
+	// we can't. We'll append new snapshots to this at the end of
+	// this func.
+	backupSnapshots, err := GetSnapshotsInBackup(backup)
+	if err != nil {
+		return err
+	}
+
+	// ensure a repo exists for the pod's namespace
+	exists, err := br.metadataManager.RepositoryExists(pod.Namespace)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		if err := br.metadataManager.InitRepo(pod.Namespace); err != nil {
+			return err
+		}
+	}
+
+	var (
+		errs []error
+		lock sync.Mutex
+		wg   sync.WaitGroup
+	)
+
+	// for each volume to backup:
+	for _, volumeName := range volumesToBackup {
+		wg.Add(1)
+		go func(volumeName string) {
+			defer wg.Done()
+
+			volume := getVolume(pod, volumeName)
+			if volume == nil {
+				lock.Lock()
```

Comment:
For coordinating goroutines & their returned errors, there's at least `golang.org/x/sync/errgroup` and @davecheney may have a better suggestion.

Reply:
Something like this would test your change: https://gist.github.com/ncdc/12c3b6b6f4db0e5c60f4ecf77e8d85c2